src/display/canvas_dependency_tracker.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * @typedef {"lineWidth" | "lineCap" | "lineJoin" | "miterLimit" | "dash" |
+ * "strokeAlpha" | "fillColor" | "fillAlpha" | "globalCompositeOperation" |
+ * "path"} SimpleDependency
+ */
+
+/**
+ * @typedef {1 | 2} Foo
+ */
+
+class CanvasDependencyTracker {
+  /** @type {Record<SimpleDependency, number>} */
+  #simple = { __proto__: null };
+
+  #incremental = {
+    __proto__: null,
+    transform: [],
+    moveText: [],
+  };
+
+  #namedDependencies = new Map();
+
+  #saves = [];
+
+  #pendingBBox = null;
+
+  #pendingBBoxIdx = null;
+
+  #pendingDependencies = new Set();
+
+  #operations = new Map();
+
+  #canvasWidth;
+
+  #canvasHeight;
+
+  constructor(initialContext) {
+    this.#canvasWidth = initialContext.canvas.width;
+    this.#canvasHeight = initialContext.canvas.height;
+  }
+
+  save(opIdx) {
+    this.#simple = { __proto__: this.#simple };
+    this.#incremental = {
+      __proto__: this.#incremental,
+      transform: { __proto__: this.#incremental.transform },
+      moveText: { __proto__: this.#incremental.moveText },
+    };
+    this.#saves.push([opIdx, null]);
+  }
+
+  restore(opIdx) {
+    this.#simple = Object.getPrototypeOf(this.#simple);
+    this.#incremental = Object.getPrototypeOf(this.#incremental);
+    this.#saves.pop()[1] = opIdx;
+  }
+
+  /**
+   * @param {SimpleDependency} name
+   * @param {number} idx
+   */
+  recordSimpleData(name, idx) {
+    this.#simple[name] = idx;
+  }
+
+  recordIncrementalData(name, idx) {
+    this.#incremental[name].push(idx);
+  }
+
+  recordNamedData(name, idx) {
+    this.#namedDependencies.set(name, idx);
+  }
+
+  resetBBox(idx) {
+    this.#pendingBBoxIdx = idx;
+    this.#pendingBBox = {
+      minX: Infinity,
+      minY: Infinity,
+      maxX: -Infinity,
+      maxY: -Infinity,
+    };
+  }
+
+  recordBBox(idx, ctx, minX, maxX, minY, maxY) {
+    const matrix = ctx.getTransform();
+
+    ({ x: minX, y: minY } = matrix.transformPoint(new DOMPoint(minX, minY)));
+    ({ x: maxX, y: maxY } = matrix.transformPoint(new DOMPoint(maxX, maxY)));
```

Comment:
Note to self: this is wrong. If we have a square and rotate it by 45deg, we'd get minX=maxX or minY=maxY.